// @storybook/angular/dist/ts3.9/client/preview/angular-beta/StorybookModule.js
"use strict";
var __makeTemplateObject = (this && this.__makeTemplateObject) || function (cooked, raw) {
    if (Object.defineProperty) { Object.defineProperty(cooked, "raw", { value: raw }); } else { cooked.raw = raw; }
    return cooked;
};
var __decorate = (this && this.__decorate) || function (decorators, target, key, desc) {
    var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
    if (typeof Reflect === "object" && typeof Reflect.decorate === "function") r = Reflect.decorate(decorators, target, key, desc);
    else for (var i = decorators.length - 1; i >= 0; i--) if (d = decorators[i]) r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
    return c > 3 && r && Object.defineProperty(target, key, r), r;
};
var __spreadArrays = (this && this.__spreadArrays) || function () {
    for (var s = 0, i = 0, il = arguments.length; i < il; i++) s += arguments[i].length;
    for (var r = Array(s), k = 0, i = 0; i < il; i++)
        for (var a = arguments[i], j = 0, jl = a.length; j < jl; j++, k++)
            r[k] = a[j];
    return r;
};
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.createStorybookModule = exports.getStorybookModuleMetadata = void 0;
var core_1 = require("@angular/core");
var platform_browser_1 = require("@angular/platform-browser");
var ts_dedent_1 = __importDefault(require("ts-dedent"));
var util_deprecate_1 = __importDefault(require("util-deprecate"));
var StorybookProvider_1 = require("./StorybookProvider");
var NgModulesAnalyzer_1 = require("./utils/NgModulesAnalyzer");
var NgComponentAnalyzer_1 = require("./utils/NgComponentAnalyzer");
var StorybookWrapperComponent_1 = require("./StorybookWrapperComponent");
var ComputesTemplateFromComponent_1 = require("./ComputesTemplateFromComponent");
var deprecatedStoryComponentWarning = util_deprecate_1.default(function () { }, ts_dedent_1.default(templateObject_1 || (templateObject_1 = __makeTemplateObject(["`component` story return value is deprecated, and will be removed in Storybook 7.0.\n        Instead, use `export const default = () => ({ component: AppComponent });`\n        or\n        ```\n        export const Primary: Story = () => ({});\n        Primary.parameters = { component: AppComponent };\n        ```\n        Read more at \n        - https://github.com/storybookjs/storybook/blob/next/MIGRATION.md#deprecated-angular-story-component).\n        - https://storybook.js.org/docs/angular/writing-stories/parameters\n      "], ["\\`component\\` story return value is deprecated, and will be removed in Storybook 7.0.\n        Instead, use \\`export const default = () => ({ component: AppComponent });\\`\n        or\n        \\`\\`\\`\n        export const Primary: Story = () => ({});\n        Primary.parameters = { component: AppComponent };\n        \\`\\`\\`\n        Read more at \n        - https://github.com/storybookjs/storybook/blob/next/MIGRATION.md#deprecated-angular-story-component).\n        - https://storybook.js.org/docs/angular/writing-stories/parameters\n      "]))));
exports.getStorybookModuleMetadata = function (_a, storyProps$) {
    var _b, _c, _d, _e, _f;
    var storyFnAngular = _a.storyFnAngular, parameters = _a.parameters;
    var storyComponent = storyFnAngular.component, props = storyFnAngular.props, styles = storyFnAngular.styles, _g = storyFnAngular.moduleMetadata, moduleMetadata = _g === void 0 ? {} : _g;
    var template = storyFnAngular.template;
    if (storyComponent) {
        deprecatedStoryComponentWarning();
    }
    var component = storyComponent !== null && storyComponent !== void 0 ? storyComponent : parameters.component;
    if (hasNoTemplate(template) && component) {
        template = ComputesTemplateFromComponent_1.computesTemplateFromComponent(component, props, '');
    }
    /**
     * Create a component that wraps generated template and gives it props
     */
    var ComponentToInject = StorybookWrapperComponent_1.createStorybookWrapperComponent(template, component, styles, props);
    // Look recursively (deep) if the component is not already declared by an import module
    var requiresComponentDeclaration = NgComponentAnalyzer_1.isDeclarable(component) &&
        !NgModulesAnalyzer_1.isComponentAlreadyDeclaredInModules(component, moduleMetadata.declarations, moduleMetadata.imports);
    return {
        declarations: __spreadArrays((requiresComponentDeclaration ? [component] : []), [
            ComponentToInject
        ], ((_b = moduleMetadata.declarations) !== null && _b !== void 0 ? _b : [])),
        imports: __spreadArrays([platform_browser_1.BrowserModule], ((_c = moduleMetadata.imports) !== null && _c !== void 0 ? _c : [])),
        providers: __spreadArrays([StorybookProvider_1.storyPropsProvider(storyProps$)], ((_d = moduleMetadata.providers) !== null && _d !== void 0 ? _d : [])),
        entryComponents: __spreadArrays(((_e = moduleMetadata.entryComponents) !== null && _e !== void 0 ? _e : [])),
        schemas: __spreadArrays(((_f = moduleMetadata.schemas) !== null && _f !== void 0 ? _f : [])),
        bootstrap: [ComponentToInject],
    };
};
exports.createStorybookModule = function (ngModule) {
    var StorybookModule = /** @class */ (function () {
        function StorybookModule() {
        }
        StorybookModule = __decorate([
            core_1.NgModule(ngModule)
        ], StorybookModule);
        return StorybookModule;
    }());
    return StorybookModule;
};
function hasNoTemplate(template) {
    return template === null || template === undefined;
}
var templateObject_1;
