// @storybook/angular/dist/ts3.9/client/preview/decorators.js
"use strict";
var __assign = (this && this.__assign) || function () {
    __assign = Object.assign || function(t) {
        for (var s, i = 1, n = arguments.length; i < n; i++) {
            s = arguments[i];
            for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p))
                t[p] = s[p];
        }
        return t;
    };
    return __assign.apply(this, arguments);
};
var __spreadArrays = (this && this.__spreadArrays) || function () {
    for (var s = 0, i = 0, il = arguments.length; i < il; i++) s += arguments[i].length;
    for (var r = Array(s), k = 0, i = 0; i < il; i++)
        for (var a = arguments[i], j = 0, jl = a.length; j < jl; j++, k++)
            r[k] = a[j];
    return r;
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.componentWrapperDecorator = exports.moduleMetadata = void 0;
var ComputesTemplateFromComponent_1 = require("./angular-beta/ComputesTemplateFromComponent");
var NgComponentAnalyzer_1 = require("./angular-beta/utils/NgComponentAnalyzer");
exports.moduleMetadata = function (metadata) { return function (storyFn) {
    var story = storyFn();
    var storyMetadata = story.moduleMetadata || {};
    metadata = metadata || {};
    return __assign(__assign({}, story), { moduleMetadata: {
            declarations: __spreadArrays((metadata.declarations || []), (storyMetadata.declarations || [])),
            entryComponents: __spreadArrays((metadata.entryComponents || []), (storyMetadata.entryComponents || [])),
            imports: __spreadArrays((metadata.imports || []), (storyMetadata.imports || [])),
            schemas: __spreadArrays((metadata.schemas || []), (storyMetadata.schemas || [])),
            providers: __spreadArrays((metadata.providers || []), (storyMetadata.providers || [])),
        } });
}; };
exports.componentWrapperDecorator = function (element, props) { return function (storyFn, storyContext) {
    var story = storyFn();
    var currentProps = typeof props === 'function' ? props(storyContext) : props;
    var template = NgComponentAnalyzer_1.isComponent(element)
        ? ComputesTemplateFromComponent_1.computesTemplateFromComponent(element, currentProps !== null && currentProps !== void 0 ? currentProps : {}, story.template)
        : element(story.template);
    return __assign(__assign(__assign({}, story), { template: template }), (currentProps || story.props
        ? {
            props: __assign(__assign({}, currentProps), story.props),
        }
        : {}));
}; };
